### PR TITLE
Change course site field date

### PR DESCRIPTION
**Pull request by MartinBentley**
### Context
There is a problem with field in the course_site table in the manage database where a date field is a string instead of a date. Here is the card:
https://trello.com/c/w7kXfqpK/872-fix-datatype-for-applicationsacceptedfrom

### Changes proposed in this pull request
The field has been changed using migrations.
There is also a data helper class to manage the conversion and tests have been added to test this new class.

### CODE REVIEW

**Comment by MartinBentley** [comment]
[Fix datatype for 'applications_accepted_from'](https://trello.com/c/w7kXfqpK/872-fix-datatype-for-applicationsacceptedfrom)

File: src/ManageCourses.UcasCourseImporter/importer/Mapping/DataHelper.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace GovUk.Education.ManageCourses.UcasCourseImporter.Mapping
+{
+    public static class DataHelper
```
Comment (timabell) [inline src/ManageCourses.UcasCourseImporter/importer/Mapping/DataHelper.cs]: how about `UcasStringParser`

File: src/ManageCourses.Api/Mapping/CourseMapper.cs [hunk 2]
```diff
@@ -114,11 +114,7 @@ public SearchAndCompare.Domain.Models.Course MapToSearchAndCompareCourse(Domain.
                     Address = address
                 },
 
-                ApplicationsAcceptedFrom = sites.Select(x =>
-                {
-                    DateTime parsed;
-                    return DateTime.TryParse(x.ApplicationsAcceptedFrom, out parsed) ? (DateTime?)parsed : null;
-                }).Where(x => x != null && x.HasValue)
+                ApplicationsAcceptedFrom = sites.Select(x => x.ApplicationsAcceptedFrom).Where(x => x != null && x.HasValue)
```
Comment (timabell) [inline src/ManageCourses.Api/Mapping/CourseMapper.cs]: redundant check

File: src/ManageCourses.UcasCourseImporter/importer/Mapping/CourseLoader.cs [hunk 3]
```diff
@@ -108,7 +108,7 @@ private Course LoadCourse(Provider provider, IEnumerable<UcasCourse> courseRecor
                 returnCourse.CourseSites = new Collection<CourseSite>(courseRecords.Select(x => new CourseSite
                 {
                     Site = allSites.Single(y => y.Provider?.ProviderCode == x.InstCode && y.Code == x.CampusCode),
-                    ApplicationsAcceptedFrom = x.CrseOpenDate,
+                    ApplicationsAcceptedFrom = DataHelper.GetDateTimeFromString(x.CrseOpenDate),
```
Comment (defong) [inline src/ManageCourses.UcasCourseImporter/importer/Mapping/CourseLoader.cs]: Can we not deal with this closer to source? Ie from the importer?

File: src/ManageCourses.Domain/Migrations/20190118140611_ChangedCourseSiteFieldDataType.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            var sql = $"ALTER TABLE \"course_site\" " +
+                      "ALTER COLUMN \"applications_accepted_from\" " +
+                      "TYPE date USING \"applications_accepted_from\"::date;";
+
+            migrationBuilder.Sql(sql);
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
+        {
+            var sql = $"ALTER TABLE \"course_site\" " +
```
[the remark this answers is not in the document]
Comment (MartinBentley) [reply src/ManageCourses.Domain/Migrations/20190118140611_ChangedCourseSiteFieldDataType.cs]: Removed Down

File: src/ManageCourses.UcasCourseImporter/importer/Mapping/CourseLoader.cs (hunk 3 above)
Comment (MartinBentley) [reply src/ManageCourses.UcasCourseImporter/importer/Mapping/CourseLoader.cs]: I considered this but rejected the idea as all the imported does is grab data from the spreadsheet. Better to put transform logic in here

**Pull request merged**